lib/Dialect/FIRRTL/Transforms/HoistPassthrough.cpp

```diff
@@ -0,0 +1,571 @@
+};
+
+} // end anonymous namespace
+
+//===----------------------------------------------------------------------===//
+// Pass Infrastructure
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct HoistPassthroughPass
+    : public HoistPassthroughBase<HoistPassthroughPass> {
+  using HoistPassthroughBase::HoistPassthroughBase;
+  void runOnOperation() override;
+
+  using HoistPassthroughBase::hoistHWDrivers;
+};
+} // end anonymous namespace
+
+void HoistPassthroughPass::runOnOperation() {
+  LLVM_DEBUG(llvm::dbgs() << "===- Running HoistPassthrough Pass "
+                             "------------------------------------------===\n");
+  auto &instanceGraph = getAnalysis<InstanceGraph>();
+
+  SmallVector<FModuleOp, 0> modules(llvm::make_filter_range(
+      llvm::map_range(
+          llvm::post_order(&instanceGraph),
+          [](auto *node) { return dyn_cast<FModuleOp>(*node->getModule()); }),
+      [](auto module) { return module; }));
+
+  MustDrivenBy driverAnalysis;
+  driverAnalysis.setIgnoreHWDrivers(!hoistHWDrivers);
+
+  // For each module (PO)...
+  for (auto module : modules) {
+    // TODO: Public means can't reason down into, or remove ports.
+    // Does not mean cannot clone out wires or optimize w.r.t its contents.
+    if (module.isPublic())
+      continue;
+
+    // 1. Analyze.
+
+    // What ports to delete.
+    // Hoisted drivers of output ports will be deleted.
+    BitVector deadPorts(module.getNumPorts());
+
+    // Analyze all ports using current IR.
+    driverAnalysis.clear();
+    driverAnalysis.run(module);
+    auto notNullAndCanHoist = [](const Driver &d) -> bool {
+      return d && d.canHoist();
+    };
+    SmallVector<Driver, 16> drivers(llvm::make_filter_range(
+        llvm::map_range(module.getArguments(),
+                        [&driverAnalysis](auto val) {
+                          return driverAnalysis.getCombinedDriverFor(val);
+                        }),
+        notNullAndCanHoist));
+
+    // 2. Rematerialize must-driven ports at instantiation sites.
+
+    // Do this first, keep alive Driver state pointing to module.
+    for (auto &driver : drivers) {
+      std::optional<size_t> deadPort;
+      {
+        auto destArg = driver.getDestBlockArg();
+        auto mod = cast<firrtl::FModuleLike>(destArg.getOwner()->getParentOp());
+        auto index = destArg.getArgNumber();
+        auto *igNode = instanceGraph.lookup(mod);
+
+        // Replace dest in all instantiations.
+        for (auto *record : igNode->uses()) {
+          auto inst = cast<InstanceOp>(record->getInstance());
+          ImplicitLocOpBuilder builder(inst.getLoc(), inst);
+          builder.setInsertionPointAfter(inst);
+
+          auto mappedDest = inst.getResult(index);
+          mappedDest.replaceAllUsesWith(driver.remat(
+              [&inst](size_t index) { return inst.getResult(index); },
+              builder));
+        }
```

Comment:
If I understand this correctly, it means
If an output port is driven by an input port, go to all instances of the module, and replace the output port with the input port ?

Reply:
Yep! Modulo and intermediate indexing or cast, but yes that's right.